### PR TITLE
Create initial OpenAPI specification

### PR DESCRIPTION
**Pull request by pxbhop**
An initial OpenAPI specification, completely open to be changed at this point in time. Initial endpoints and schemas are based on discussions within the new-api channel on Discord. Unfortunately GitHub doesn't have any built in OpenAPI doc rendering (unlike GitLab) so anyone wanting to contribute to discussion will have to look at rendering it themselves or just reading the raw YAML.

### CODE REVIEW

File: openapi.yml [hunk 1]
```diff
@@ -0,0 +1,381 @@
+openapi: '3.0.2'
```
Comment (0o22) [inline openapi.yml]: > String does not match the pattern of "^3\.1\.\d+(-.+)?$".yaml-schema: openapi.json
Source: [schema.json](https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/schemas/v3.1/schema.json)

**Review by 0o22 — CHANGES_REQUESTED** [review]

**Comment by tetsomina** [comment]
Just a friendly reminder: frick px.

File: openapi.yml [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+        '200':
+          description: A Map opject
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Map'
+        '400':
+          $ref: '#/components/responses/Invalid'
+        '401':
+          $ref: '#/components/responses/Unauthorized'
```
Comment (2xrJohann) [inline openapi.yml]: is this not gonna be public? prob dont need anything other than 200, 400 & 404

File: openapi.yml [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+      responses:
+        '200':
+          description: An array of Server objects
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ServerArray'
+        '400':
+          $ref: '#/components/responses/Invalid'
+        '401':
```
Comment (2xrJohann) [inline openapi.yml]: feels like get /servers shouldnt require auth

File: openapi.yml [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+          description: Unexpected error
+
+# We can store common response objects and types here
+components:
+  # Security schemes for protected endpoints (run submission)
+  securitySchemes:
+    ApiKeyAuth:
+      type: apiKey
+      in: header
+      name: X-API-Key
```
Comment (2xrJohann) [inline openapi.yml:288]: iirc there were discussions of using next-auth, might be better to run w/ JWT in this case

**Review by 2xrJohann — COMMENTED** [review]
feel free to resolve threads if you want, just my initial thoughts

File: openapi.yml (hunk 2 above)
Comment (pxbhop) [reply openapi.yml]: lowkey copy/pasted the error responses across all paths, forgot to remove 401 from everything but /runs POST

File: openapi.yml (hunk 3 above)
Comment (pxbhop) [reply openapi.yml]: see other comment about 401 on maps

File: openapi.yml (hunk 4 above)
Comment (pxbhop) [reply openapi.yml:288]: atm everything relating to runs /POST is just for reference while the api is written, it'll be removed from docs later, endpoint will require an api key to access from the server plugin (at this point in time) - next-auth would be used for site login stuff i imagine

File: openapi.yml (hunk 2 above)
Comment (2xrJohann) [reply openapi.yml]: feels just like being employed again ✊

File: openapi.yml (hunk 1 above)
Comment (pxbhop) [reply openapi.yml]: OpenAPI 3.0.2 was still a valid version, I've updated to 3.1.0 now though

File: openapi.yml [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+      schema:
+        type: string
+        enum:
+          - asc
+          - desc
+      description: >
+        Sort order (by Run ID):
+          * `asc` - Ascending, from low to high
+          * `desc` - Descending, from high to low
+    LimitParam:
```
Comment (2xrJohann) [inline openapi.yml:588]: imo limit should have a sensible default and maximum, prob minimum of 1 too

Comment (2xrJohann) [reply openapi.yml:588]: and prob page number on apis that use it or something similar to traverse further

Comment (pxbhop) [reply openapi.yml:588]: default and min/max limit sounds fine, idk if we need page number if we have offset tho?

Comment (2xrJohann) [reply openapi.yml:588]: ye dont need w/ offset 🚀

**Comment by ecsr** [comment]
Thanks!

**Pull request merged**